File: src/main/scala/homework/HW1.scala
package homework

object HW1 {

  trait Exp
  case class StringLit(s:String) extends Exp
  case class Concat(left: Exp, right: Exp) extends Exp
  case class RestAfter(left: Exp, right: Exp) extends Exp

  object HW1Interpreter extends Interpreter[Exp, String] with Parser[Exp] {
    def eval(exp: Exp): String = {
      exp match {
        case StringLit(s) => s.drop(1).dropRight(1)
        case Concat(l, r) => eval(l) + eval(r)
        case RestAfter(l, r) => {
          val ls = eval(l)
          val rs = eval(r)
          val i = ls indexOf rs
          if (i == -1) error(rs + " not in " + ls) else ls.substring(i+1)
        }
      }
    }

    def parse(expr: Any): Exp = {
      expr match {
        case s: String => StringLit(s)
        case List(xs, Symbol("&"), ys) => Concat(parse(xs), parse(ys))
        case List(xs, Symbol("@"), ys) => RestAfter(parse(xs), parse(ys))
        case _ => error("unexpected token: " + expr)
      }
    }
  }
}

File: src/main/clojure/hw1.clj
(ns hw1 
  (:use clojure.test))

(defrecord StringLit [s])
(defrecord Concat [l r])
(defrecord RestAfter [l r])

(defn parse [sexp]
  (cond 
    (symbol? sexp) (StringLit. (str sexp))
    (list? sexp)
      (let [[left op right] sexp]
        (case op
          & (Concat. (parse left) (parse right))
          % (RestAfter. (parse left) (parse right))
          :else (throw (IllegalArgumentException. (str "bad list: " sexp))))) 
    :else (throw (IllegalArgumentException. (str "bad arg: " sexp)))))
 
(defmulti interpret class)
(defmethod interpret StringLit [s] (:s s))  
(defmethod interpret Concat [c] (str (interpret (:l c)) (interpret (:r c))))  
(defmethod interpret RestAfter [ra]
  (let [l (interpret (:l ra)) r (interpret (:r ra)) i (.indexOf l r)]
    (if (< i 0) (throw (Exception. (str r " not found in " l)))
      		(.substring l (+ i (.length r))))))

(defn testcode [l r]
  (with-test (defn f [x] l) (is (= (interpret (parse l)) r)))
)
(testcode 'hello "hello") 
(testcode '((hello & x) & world) "helloxworld")
(testcode '((hello % l) & (aworld % a)) "loworld")
(testcode '(((a & b) & (a & b)) & ((a & b) & (a & b))) "abababab")
(testcode '((((((abababab % a) % b) % a) % b) % a) % b) "ab")
(testcode '(((((((a & b) & (a & b)) & ((a & b) & (a & b))) % ab) % ab) % ab) % ab) "")
(testcode '(filename.scm % .) "scm")

(run-tests)
